Rotation constraint bilinear

hongkai-dai opened the pull request:
This introduces a different mixed-integer convex relaxation on the SO(3) rotation constraint. The idea is to replace the bilinear product in the SO(3) constraint, with an auxiliary variable lies in the convex hull of the curve w = x * y. We test this new relaxation, and found it being much tighter than the one that was in the master, if we use 1 interval per half axis, and slightly tighter with 2 intervals per half axis.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6635)
<!-- Reviewable:end -->

hongkai-dai commented:
+@gizatt for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpbUf02UF15LsqCpaHX:biagjkq)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
@drake-jenkins-bot mac-clang-bazel-experimental please.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpbcPNCrdPKTB5qPRQb:b-mdoehw)*
<!-- Sent from Reviewable.io -->

gizatt commented:
Reviewed 1 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/solvers/rotation_constraint.cc, line 893 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppitifMajir9uTo0Bs:-KppitigUWm5aeY3INCN:burk497) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L893)):*
> ```cpp
> }
> 
> // Relax the unit length constraint x₀² + x₁² + x₂² = 1, to a set of linear
> ```

Given that you supply \phi and \lambda to this function (and not x), I think the documentation would be clearer that this function constructs x out of \phi and \lambda according to the below rule (and not that x should be passed...). It caught me up for a few minutes

---

*[drake/solvers/rotation_constraint.cc, line 901 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppiBelQxTt41yh-bpn:-KppiBelQxTt41yh-bpo:bgmbgmx) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L901)):*
> ```cpp
> // w = x² at φⱼ. Thus we have 1 ≥ sum_j sum_i 2φⱼ*xᵢ-φⱼ²
> // Moreover, also due to the convexity of the curve w = x², we know
> //   x² ≤ sum_j φ(j)² * λ(j)
> ```

Should this be x_i and \lambda_i (j), like below?

---

*[drake/solvers/rotation_constraint.cc, line 904 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppknMRQaQUHMtvUao0:-KppknMRQaQUHMtvUao1:b-ihlt6s) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L904)):*
> ```cpp
> // So we have the constraint
> // 1 ≤ sum_i sum_j φ(j)² * λᵢ(j)
> void AddUnitLengthConstraintWithLogarithmicSos2(
> ```

Does this have to be logarithmic SOS2? Seems like it ought to work no matter what SOS2 implementation is enforced on \lambda.

---

*[drake/solvers/rotation_constraint.cc, line 904 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpplGEm-MHD-myreo28:-KpplGEm-MHD-myreo29:b-6ypnck) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L904)):*
> ```cpp
> // So we have the constraint
> // 1 ≤ sum_i sum_j φ(j)² * λᵢ(j)
> void AddUnitLengthConstraintWithLogarithmicSos2(
> ```

Might be a tiny bit more correct to be "AddUnitLengthConstraintTo[Logarithmic?]Sos2Variables" or something to imply that this SOS2 constraint is assumed, not applied, by this function

---

*[drake/solvers/rotation_constraint.cc, line 987 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kppokch2_Tb8-p2b-Iq:-Kppokch2_Tb8-p2b-Ir:b-kx2yui) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L987)):*
> ```cpp
>           prog, R(Ri_row, Ri_col), R(Rj_row, Rj_col), W(i, j), phi, phi,
>           B[Ri_row][Ri_col], B[Rj_row][Rj_col]);
>       prog->AddLinearConstraint(
> ```

Could you explain a bit about what's going on with \lambda here?

---

*[drake/solvers/rotation_constraint.cc, line 1001 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpporWhGW6cDQAoiGWf:-KpporWhGW6cDQAoiGWg:bb4zmv) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1001)):*
> ```cpp
>   for (int i = 0; i < 3; ++i) {
>     for (int j = i + 1; j < 3; ++j) {
>       prog->AddLinearConstraint(
> ```

Comment that these are the dot product orthogonality constraints?

---

*[drake/solvers/rotation_constraint.cc, line 1016 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppowHv9_6vqQbmC1Xb:-KppowHv9_6vqQbmC1Xc:b-yol3rv) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1016)):*
> ```cpp
>     int j = (i + 1) % 3;
>     int k = (i + 2) % 3;
>     Vector3<symbolic::Expression> cross_product1 =
> ```

Comment that these are the cross product constraints?

---

*[drake/solvers/rotation_constraint.cc, line 1160 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kppq7k7AZelpxwfEOZQ:-Kppq7k7AZelpxwfEOZR:bxolhz4) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1160)):*
> ```cpp
>     MathematicalProgram *prog,
>     const Eigen::Ref<const MatrixDecisionVariable<3, 3>> &R,
>     int num_intervlas_per_half_axis) {
> ```

intervlas -> intervals

---

*[drake/solvers/rotation_constraint.cc, line 1187 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpprQZL4TyCgp3Jj7fJ:-KpprQZL4TyCgp3Jj7fK:b-c1vllq) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1187)):*
> ```cpp
>     for (int j = 0; j < 3; ++j) {
>       lambda[i][j] = prog->NewContinuousVariables<kLambdaRows, 1>(lambda_rows,
>                                                                   1, "lambda");
> ```

should this name have a unique identifier in it to indicate that it's lambda_[i, j]?

---

*[drake/solvers/rotation_constraint.h, line 184 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppqqKYXhmF8M7W9N9d:-KppqqKYXhmF8M7W9N9e:b-z9d9xe) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.h#L184)):*
> ```objc
>  * @param prog The optimization program to which the SO(3) relaxation is added.
>  * @param R The rotation matrix.
>  * @param num_intervlas_per_half_axis Same as NumIntervalsPerHalfAxis, use this
> ```

intervlas->intervals

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 2 of 5 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/solvers/rotation_constraint.cc, line 901 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppiBelQxTt41yh-bpn:-Kppw5DTgzTllcxwkiTy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L901)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Should this be x_i and \lambda_i (j), like below?
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 904 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppknMRQaQUHMtvUao0:-KppwC6K6OG3vj3ykwNm:b-h5g1tk) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L904)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Does this have to be logarithmic SOS2? Seems like it ought to work no matter what SOS2 implementation is enforced on \lambda.
</blockquote></details>

Done. Good call

---

*[drake/solvers/rotation_constraint.cc, line 904 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpplGEm-MHD-myreo28:-KppxcXszbRV5SCgX-U1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L904)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Might be a tiny bit more correct to be "AddUnitLengthConstraintTo[Logarithmic?]Sos2Variables" or something to imply that this SOS2 constraint is assumed, not applied, by this function
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 987 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kppokch2_Tb8-p2b-Iq:-Kpq-D76mOPRZhpZ7137:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L987)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Could you explain a bit about what's going on with \lambda here?
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 1001 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpporWhGW6cDQAoiGWf:-Kpq-qZFrCdbz9z7z-eF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1001)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Comment that these are the dot product orthogonality constraints?
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 1016 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppowHv9_6vqQbmC1Xb:-Kpq038e7wyxs2Xf8HZQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1016)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Comment that these are the cross product constraints?
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 1160 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kppq7k7AZelpxwfEOZQ:-Kpq09DIs6U3De5WZTTx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1160)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

intervlas -> intervals
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 1187 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpprQZL4TyCgp3Jj7fJ:-Kpq0RRqbUzRa34GvDTY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L1187)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

should this name have a unique identifier in it to indicate that it's lambda_[i, j]?
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.h, line 184 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppqqKYXhmF8M7W9N9d:-Kpq0_VPwSUpD_8Sckn0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.h#L184)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

intervlas->intervals
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635)*
<!-- Sent from Reviewable.io -->

gizatt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/rotation_constraint.cc, line 893 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KppitifMajir9uTo0Bs:-KpqCogXUpDwqvxkZ1vt:b-b246a7) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L893)):*
<details><summary><i>Previously, gizatt (Gregory Izatt) wrote…</i></summary><blockquote>

Given that you supply \phi and \lambda to this function (and not x), I think the documentation would be clearer that this function constructs x out of \phi and \lambda according to the below rule (and not that x should be passed...). It caught me up for a few minutes
</blockquote></details>

That helps a lot!

---

*[drake/solvers/rotation_constraint.cc, line 987 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kppokch2_Tb8-p2b-Iq:-KpqD5DOz0ZY5ZaTwDef:b15qme6) ([raw file](https://github.com/robotlocomotion/drake/blob/177ecb7892149e8d39517b527d1ad6c330690f5f/drake/solvers/rotation_constraint.cc#L987)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!  That helps a lot

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpqDkSGGJ15sKV-u6PB:bnfp4nl)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+@jwnimmer-tri for platform review please.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpqFvB0ohvhSVQgcNcv:b-qxjhdt)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging) 

---

Reviewed 1 of 5 files at r1, 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/CMakeLists.txt, line 136 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqjQfq6_nIaRvkuzhG:-KpqjQfrOqMVdEY30upc:b-wdnfzv) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/CMakeLists.txt#L136)):*
> ```CMake
>         LIBS -ldrakeBilinearProductUtil
>         REQUIRES
>         drake-common)
> ```

BTW I think this requires drake-optimiztation because it uses decision_variable?

---

*[drake/solvers/rotation_constraint.cc, line 908 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqqpbvyfnb_iNjUncO:-Kpqqpbvyfnb_iNjUncP:b8bjg0d) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L908)):*
> ```cpp
> // 1 ≤ sum_i sum_j φ(j)² * λᵢ(j)
> void AddUnitLengthConstraintWithSos2Lambda(
>     MathematicalProgram *prog, const Eigen::Ref<const Eigen::VectorXd> &phi,
> ```

_Meta_ the `&` and `*` should cuddle with the typename, not the variable name.

---

*[drake/solvers/rotation_constraint.cc, line 945 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqqaG0pVbTijNvWwqO:-KpqqaG0pVbTijNvWwqP:b6p4tim) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L945)):*
> ```cpp
> std::pair<int, int> Index2Subscripts(int index, int num_rows, int num_cols) {
>   DRAKE_ASSERT(index >= 0 && index < num_rows * num_cols);
>   int col_idx = index / num_rows;
> ```

BTW nit _"... do not abbreviate by deleting letters within a word."_
-- https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

---

*[drake/solvers/rotation_constraint.cc, line 1185 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqpHtLXnfccSJWhXnh:-KpqpHtLXnfccSJWhXni:bci1ly9) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L1185)):*
> ```cpp
>       typename AddRotationMatrixBilinearMcCormickMilpConstraintsReturn<
>           NumIntervalsPerHalfAxis>::PhiType PhiType;
>   DRAKE_DEMAND(num_intervals_per_half_axis >= 1);
> ```

FYI Assuming the change to use overloading does not happen, consider asserting that `num_intervals_per_half_axis` is correct (matches) when `NumIntervalsPerHalfAxis >= 1`.

---

*[drake/solvers/rotation_constraint.h, line 159 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqlPis0-qb5NOIRV_C:-KpqlPis0-qb5NOIRV_D:b-x9v4mb) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L159)):*
> ```objc
>                                      : 1 + 2 * NumIntervalsPerHalfAxis;
>   static constexpr int NumBinaryVars =
>       PhiRows == Eigen::Dynamic ? Eigen::Dynamic : CeilLog2(PhiRows - 1);
> ```

https://google.github.io/styleguide/cppguide.html#Constant_Names is pretty clear that this should be kFoo and not just Foo, because we are using static storage duration.

I think there is a fair argument for "stay in sync with Eigen" yielding an exception to the rule for cases like `LogarithmicSos2NewBinaryVariables::Rows` because its literally meant to be a `Rows` argument to Eigen.  These are also, but aren't called just `Rows` but are `FooRows` or even `NumFooVars` which is used as a count of rows.

It is definitely a readability burden to remember that SomeThing is a int and not a class, but perhaps the burden of kSomeThing is worse in some cases?

If we indeed are going to have so many cases where we have integral constants in template programming, perhaps it is worth taking a leap now and doing `kFoo` pervasively in the solvers code?

---

*[drake/solvers/rotation_constraint.h, line 164 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqjhH4qiZuAkFAsPOp:-KpqjhH4qiZuAkFAsPOq:b-ucrg82) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L164)):*
> ```objc
>   typedef Eigen::Matrix<double, PhiRows, 1> PhiType;
>   typedef std::pair<BinaryVarType, PhiType> type;
> };
> ```

This struct lacks Doxygen documentation (class overview, at a minimum).

I suspect that's because its an internal implementation detail and should not be something users care about.  If so, it should be in a `detail` namespace to hide it, at which point it doesn't require Doxygen.

---

*[drake/solvers/rotation_constraint.h, line 190 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqoZJFSI8ajPJ1_aON:-KpqoZJFSI8ajPJ1_aOO:b-38avk0) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L190)):*
> ```objc
>  * interval [φ(M), φ(M+1)]. φ contains the end points of the all the intervals,
>  * namely φ(i) = -1 + 1 / num_intervals_per_half_axis * i.
>  */
> ```

The API docs should explain that only certain values of NumIntervalsPerHalfAxis are supported (Dynamic, and 1..4).

FYI Also consider changing the enable_if so that if someone uses `5` they would get a compiler error instead of a linker error.

---

*[drake/solvers/rotation_constraint.h, line 198 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqmaxN3xk8lBz4-HR6:-KpqmaxN3xk8lBz4-HR7:b-ubncz9) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L198)):*
> ```objc
> AddRotationMatrixBilinearMcCormickMilpConstraints(
>     MathematicalProgram *prog,
>     const Eigen::Ref<const MatrixDecisionVariable<3, 3>> &R,
> ```

BTW nit The `&` should cuddle with the typename, not the variable name.
Ditto for the `*` in the line prior.

Ditto for the `.cc` file.

---

*[drake/solvers/rotation_constraint.h, line 199 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqmt0NVj3wolu_46hE:-Kpqmt0NVj3wolu_46hF:be528w9) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L199)):*
> ```objc
>     MathematicalProgram *prog,
>     const Eigen::Ref<const MatrixDecisionVariable<3, 3>> &R,
>     int num_intervals_per_half_axis = NumIntervalsPerHalfAxis);
> ```

FYI If I understand correctly, if `NumInter... == Dynamic`, then `num_inter...` must not be `Dynamic` but instead must be a non-negative runtime size must be passed in.  Is that correct?

Assuming so, then consider overloading this method instead of having a disjunction (`||`) in the `enable_if` -- so that callers with a fixed size _must not_ specify the `num_inter...`, and callers with a dynamic size _must always_ specify the `num_inter...`.  They could still delegate to the same implementation internally.

---

*[drake/solvers/test/rotation_constraint_test.cc, line 262 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqsOF5fh1GYMwgPX-v:-KpqsOF6h2Y7kwCKvPWj:b-mypupo) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L262)):*
> ```cpp
>   // another varaible in the McCormick envelope. Otherwise, relax the surface of
>   // the unit sphere to its convex hull.
>   int num_intervals_per_half_axis_;
> ```

Missing `{}` on non-const POD member field.

---

*[drake/solvers/test/rotation_constraint_test.cc, line 268 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqt13WOJJJZCh00CG2:-Kpqt13WOJJJZCh00CG3:b-5xx29a) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L268)):*
> ```cpp
> TEST_P(TestMcCormick, TestExactRotationMatrix) {
>   // If R is exactly on SO(3), test whether it also satisfies our relaxation.
>   std::mt19937 generator(41);
> ```

Declare local variables as close as practical to their point of use.  (Move this lower.)

---

*[drake/solvers/test/rotation_constraint_test.cc, line 270 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqt53cKOwn_79N08_S:-Kpqt53cKOwn_79N08_T:btq0qw3) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L270)):*
> ```cpp
>   std::mt19937 generator(41);
>   std::normal_distribution<double> randn;
>   std::uniform_int_distribution<> rand(0, 1 << 6);
> ```

BTW nit The prior two variables are unused.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpqiQGEdsZVRzCJnTlW:b-s794vx)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/solvers/CMakeLists.txt, line 136 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqjQfq6_nIaRvkuzhG:-KprgUg5VWqBWgLd0-3n:b-h5g1tk) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/CMakeLists.txt#L136)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I think this requires drake-optimiztation because it uses decision_variable?
</blockquote></details>

Done. Good call

---

*[drake/solvers/rotation_constraint.cc, line 908 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqqpbvyfnb_iNjUncO:-Kprgi_VmSFQtLxh0jUu:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L908)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ the `&` and `*` should cuddle with the typename, not the variable name.
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 945 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqqaG0pVbTijNvWwqO:-KprgrfzByR3x4dKhdBe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L945)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit _"... do not abbreviate by deleting letters within a word."_
-- https://google.github.io/styleguide/cppguide.html#General_Naming_Rules
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.cc, line 1185 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqpHtLXnfccSJWhXnh:-KprhAtICosBQK_7wsZI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.cc#L1185)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Assuming the change to use overloading does not happen, consider asserting that `num_intervals_per_half_axis` is correct (matches) when `NumIntervalsPerHalfAxis >= 1`.
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.h, line 159 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqlPis0-qb5NOIRV_C:-KprhpYeHTHtS5xGBeaY:b5ds3gh) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L159)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

https://google.github.io/styleguide/cppguide.html#Constant_Names is pretty clear that this should be kFoo and not just Foo, because we are using static storage duration.

I think there is a fair argument for "stay in sync with Eigen" yielding an exception to the rule for cases like `LogarithmicSos2NewBinaryVariables::Rows` because its literally meant to be a `Rows` argument to Eigen.  These are also, but aren't called just `Rows` but are `FooRows` or even `NumFooVars` which is used as a count of rows.

It is definitely a readability burden to remember that SomeThing is a int and not a class, but perhaps the burden of kSomeThing is worse in some cases?

If we indeed are going to have so many cases where we have integral constants in template programming, perhaps it is worth taking a leap now and doing `kFoo` pervasively in the solvers code?
</blockquote></details>

Done. I changed them all to `kFoo` for the code in this PR. Will do a separate PR to change `Foo` to `kFoo` in other files in solvers code.

---

*[drake/solvers/rotation_constraint.h, line 164 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqjhH4qiZuAkFAsPOp:-KpridDcuhxnTHyMNnlE:bf7rrl4) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L164)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This struct lacks Doxygen documentation (class overview, at a minimum).

I suspect that's because its an internal implementation detail and should not be something users care about.  If so, it should be in a `detail` namespace to hide it, at which point it doesn't require Doxygen.
</blockquote></details>

Done. I added documentation here.

---

*[drake/solvers/rotation_constraint.h, line 190 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqoZJFSI8ajPJ1_aON:-KprjKaoMnG0KEhK85h-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L190)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The API docs should explain that only certain values of NumIntervalsPerHalfAxis are supported (Dynamic, and 1..4).

FYI Also consider changing the enable_if so that if someone uses `5` they would get a compiler error instead of a linker error.
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.h, line 198 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqmaxN3xk8lBz4-HR6:-KprjV_taNfJiscsB1V3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L198)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit The `&` should cuddle with the typename, not the variable name.
Ditto for the `*` in the line prior.

Ditto for the `.cc` file.
</blockquote></details>

Done.

---

*[drake/solvers/rotation_constraint.h, line 199 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqmt0NVj3wolu_46hE:-KprjlIt3q0AXKOCpfgf:b-5lraky) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/rotation_constraint.h#L199)):*
> FYI If I understand correctly, if NumInter... == Dynamic, then num_inter... must not be Dynamic but instead must be a non-negative runtime size must be passed in. Is that correct?

That is correct.

I actually want to allow the user to specify the number of intervals per half axis with the same API, for both compile time value, or runtime value. So that the following code can work
```C++
template<typename Derived>
void foo(const Derived& x) {
  MathematicalProgram prog;
  auto R = prog.NewContinuousVariables<3, 3>();
  AddRotationMatrixBilinearMcCormickMilpConstraints<Derived::RowsAtCompileTime>(&prog, R, x.rows());
}
```

In `rotation_constraint.cc`, we already have the check
```C++
DRAKE_DEMAND(num_intervals_per_half_axis == kNumIntervalsPerHalfAxis ||
               kNumIntervalsPerHalfAxis == Eigen::Dynamic);
DRAKE_DEMAND(num_intervals_per_half_axis >= 1);
```
So the user should receive an error in the wrong time, if they specified both `kNumIntervalsPerHalfAxis` as a positive integer, and `num_intervals_per_half_axis` not equal to `kNumIntervalsPerHalfAxis`.

I tried to keep this API, while adding the overloaded API for both static and dynamic sized case, but the linker fails, that it says I have multiple definition. The code is in https://github.com/hongkai-dai/drake/blob/rotation_constraint_bilinear_overload/drake/solvers/rotation_constraint.h#L212-L223

---

*[drake/solvers/test/rotation_constraint_test.cc, line 262 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-KpqsOF5fh1GYMwgPX-v:-KprnkrysYTV7rZI6LjI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L262)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing `{}` on non-const POD member field.
</blockquote></details>

Done.

---

*[drake/solvers/test/rotation_constraint_test.cc, line 268 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqt13WOJJJZCh00CG2:-KprnreyhfHGeefmSzWI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L268)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Declare local variables as close as practical to their point of use.  (Move this lower.)
</blockquote></details>

Done.

---

*[drake/solvers/test/rotation_constraint_test.cc, line 270 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-Kpqt53cKOwn_79N08_S:-Kprns5a2WguOZDBpQff:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/028451a87850266f4dcd2a57917ee936c4563f9e/drake/solvers/test/rotation_constraint_test.cc#L270)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit The prior two variables are unused.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KptRam7V3W1JxPgDVB-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6635#-:-KpuP41z3xvqIyJ4-GvF:b-l70wzf)*
<!-- Sent from Reviewable.io -->
